src/python/pants/backend/experimental/python/lockfile.py

```diff
@@ -53,28 +54,54 @@ class LockGoal(Goal):
     subsystem_cls = LockSubsystem
 
 
-# TODO: What should the input to `./pants lock` be when generating user lockfiles? Currently, it's
-#  the 3rd-party requirements that get used as input. Likely it should instead be all 3rd party
-#  reqs used by the transitive closure? That would better mirror generate_lockfile.sh.
 @goal_rule
 async def generate_lockfile(
-    targets: Targets, pip_tools_subsystem: PipToolsSubsystem, workspace: Workspace
+    addresses: Addresses, pip_tools_subsystem: PipToolsSubsystem, workspace: Workspace
 ) -> LockGoal:
+    # TODO: Looking at the transitive closure to generate a single lockfile will not work when we
+    #  have multiple lockfiles supported, via per-tool lockfiles and multiple user lockfiles.
+    #  Ideally, `./pants lock ::` would mean "regenerate all unique lockfiles", whereas now it
+    #  means "generate a single lockfile based on this transitive closure."
+    transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest(addresses))
     reqs = PexRequirements.create_from_requirement_fields(
-        tgt[PythonRequirementsField] for tgt in targets if tgt.has_field(PythonRequirementsField)
+        tgt[PythonRequirementsField]
+        # NB: By looking at the dependencies, rather than the closure, we only generate for
+        # requirements that are actually used in the project.
+        #
+        # TODO: It's not totally clear to me if that is desirable. Consider requirements like
+        #  pydevd-pycharm. Should that be in the lockfile? I think this needs to be the case when
+        #  we have multiple lockfiles, though: we shouldn't look at the universe in that case,
+        #  only the relevant subset of requirements.
```

Comment:
Not sure if it's what you're getting at here, but: the lockfile(s) == "the resolve": if when running/testing or packaging the code we would include a dep, it must be in the lockfile, because when the lockfile is unchanged, we will use the lockfile as the key that identifies an invoke of PEX to build the resolve (which will then be subsetted).

Reply:
pydevd-pycharm is an interesting one because it's not normally used in tests, only when Pants developers add it during debugging. This would mean that they need to regenerate the lockfile every time they add the import. I think that's expected, right?

Reply:
Hm. Yes, unless perhaps it was "forced" into the resolve by some `*_binary`/`*_test` declaration (hand-wave, but you could do it explicitly)? That way it would always be in the resolve, and would just be filtered out by subsetting. Adding it to a test would then not need to re-run the resolve.

Reply:
Yeah, that makes sense. The bigger thing I'm after is that Pants _should_ only create a lockfile for reqs that are actually used in your project.

Even if you requirements.txt has `some-req`, if it's never used anywhere, I think we should skip it in the lockfile. And I think that is going to be necessary when we have multiple user lockfiles because it might not be safe to include some of those global, unused reqs in that resolve.

Reply:
> Even if you requirements.txt has `some-req`, if it's never used anywhere, I think we should skip it in the lockfile. And I think that is going to be necessary when we have multiple user lockfiles because it might not be safe to include some of those global, unused reqs in that resolve.

Hm. Yes, but ... lockfile == resolve. It's already the case that if you have a `requirements.txt` file containing unused stuff, the unused stuff will not actually be included in any of the resolves in your repository

... _except_ when "use all constraints" is set. That setting does something different than what a "named resolve" concept would do. Each named resolve contains only the subset of requirements _actually used_ by the roots participating in the resolve.